pkg/controllers/machine/termination/controller.go

```diff
@@ -0,0 +1,138 @@
+	cloudProvider cloudprovider.CloudProvider
+	recorder      events.Recorder
+	terminator    *terminator.Terminator
+}
+
+// NewController is a constructor for the Machine Controller
+func NewController(kubeClient client.Client, cloudProvider cloudprovider.CloudProvider,
+	terminator *terminator.Terminator, recorder events.Recorder) corecontroller.Controller {
+	return corecontroller.Typed[*v1alpha5.Machine](kubeClient, &Controller{
+		kubeClient:    kubeClient,
+		cloudProvider: cloudProvider,
+		recorder:      recorder,
+		terminator:    terminator,
+	})
+}
+
+func (*Controller) Name() string {
+	return "machine_termination"
+}
+
+func (c *Controller) Reconcile(_ context.Context, _ *v1alpha5.Machine) (reconcile.Result, error) {
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Finalize(ctx context.Context, machine *v1alpha5.Machine) (reconcile.Result, error) {
+	stored := machine.DeepCopy()
+	if !controllerutil.ContainsFinalizer(machine, v1alpha5.TerminationFinalizer) {
+		return reconcile.Result{}, nil
+	}
+	if err := c.cleanupNodeForMachine(ctx, machine); err != nil {
+		if terminator.IsNodeDrainError(err) {
+			return reconcile.Result{Requeue: true}, nil
+		}
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).With("provisioner", machine.Labels[v1alpha5.ProvisionerNameLabelKey], "provider-id", machine.Status.ProviderID))
+	if machine.Status.ProviderID != "" {
+		if err := c.cloudProvider.Delete(ctx, machine); cloudprovider.IgnoreMachineNotFoundError(err) != nil {
+			return reconcile.Result{}, fmt.Errorf("terminating cloudprovider instance, %w", err)
+		}
+	}
+	controllerutil.RemoveFinalizer(machine, v1alpha5.TerminationFinalizer)
+	if !equality.Semantic.DeepEqual(stored, machine) {
+		if err := c.kubeClient.Patch(ctx, machine, client.MergeFrom(stored)); err != nil {
+			return reconcile.Result{}, client.IgnoreNotFound(fmt.Errorf("removing machine termination finalizer, %w", err))
+		}
+		logging.FromContext(ctx).Infof("deleted machine")
+	}
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) cleanupNodeForMachine(ctx context.Context, machine *v1alpha5.Machine) error {
+	node, err := machineutil.NodeForMachine(ctx, c.kubeClient, machine)
+	if err != nil {
+		// We don't clean the node if we either don't find a node or have violated the single machine to single node invariant
+		if machineutil.IsNodeNotFoundError(err) || machineutil.IsDuplicateNodeError(err) {
+			return nil
+		}
+		return err
+	}
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).With("node", node.Name))
+	if err = c.terminator.Cordon(ctx, node); err != nil {
+		return fmt.Errorf("cordoning node, %w", err)
+	}
+	if err = c.terminator.Drain(ctx, node); err != nil {
+		if terminator.IsNodeDrainError(err) {
+			c.recorder.Publish(terminatorevents.NodeFailedToDrain(node, err))
+		}
+		return fmt.Errorf("draining node, %w", err)
+	}
+	// Best effort to remove the finalizer from the Node
```

Comment:
It seems wrong to not handle the error on the patch. What's a reason why removing the finalizer could fail, but we want to continue anyway?

Reply:
This is meant to short-circuit the deletion flow. The problem is that if we fail to patch and then we requeue, we're just making extra launch calls for no reason and slowing everything down

Reply:
Doesn't this termination controller act independent of the other machine controllers? If we requeue, we're just going to be requeuing into this controller, which doesn't have any launch calls. I think I'm misunderstanding how this all fits together because the code isn't active yet.